src/connector/src/sink/file_sink/mod.rs

```diff
@@ -0,0 +1,230 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+pub mod fs;
+pub mod gcs;
+pub mod opendal_sink;
+pub mod s3;
+use std::collections::HashMap;
+use std::sync::Arc;
+
+use anyhow::anyhow;
+use arrow_schema::{DataType as ArrowDataType, Field as ArrowField, SchemaRef};
+use async_trait::async_trait;
+use opendal::{Operator, Writer as OpendalWriter};
+use parquet::arrow::AsyncArrowWriter;
+use parquet::file::properties::WriterProperties;
+use risingwave_common::array::{to_record_batch_with_schema, Op, StreamChunk};
+use risingwave_common::buffer::Bitmap;
+use risingwave_common::catalog::Schema;
+
+use crate::sink::catalog::SinkEncode;
+use crate::sink::{Result, SinkError, SinkWriter};
+
+// Todo(wcy-fdu): upgrade to opendal 0.47.
+const SINK_WRITE_BUFFER_SIZE: usize = 16 * 1024 * 1024;
+
+pub struct OpenDalSinkWriter {
+    schema: SchemaRef,
+    operator: Operator,
+    sink_writer: Option<FileWriterEnum>,
+    pk_indices: Vec<usize>,
+    is_append_only: bool,
+    write_path: String,
+    epoch: Option<u64>,
+    executor_id: u64,
+    encode_type: SinkEncode,
+}
+
+/// The `FileWriterEnum` enum represents different types of file writers used for various sink
+/// implementations.
+///
+/// # Variants
+///
+/// - `ParquetFileWriter`: Represents a Parquet file writer using the `AsyncArrowWriter<W>`
+/// for writing data to a Parquet file. It accepts an implementation of W: `AsyncWrite` + `Unpin` + `Send`
+/// as the underlying writer. In this case, the `OpendalWriter` serves as the underlying writer.
+///
+/// - `FileWriter`: Represents a file writer for sinks other than Parquet. It uses the `OpendalWriter`
+/// directly for writing data to the file.
+///
+/// The choice of writer used during the actual writing process depends on the encode type of the sink.
+enum FileWriterEnum {
+    ParquetFileWriter(AsyncArrowWriter<OpendalWriter>),
+    FileWriter(OpendalWriter),
+}
+
+#[async_trait]
+impl SinkWriter for OpenDalSinkWriter {
+    async fn write_batch(&mut self, chunk: StreamChunk) -> Result<()> {
+        // Note: epoch is used to name the output files.
+        // Todo: after enabling sink decouple, use the new naming convention.
+        let epoch = self.epoch.ok_or_else(|| {
+            SinkError::Opendal("epoch has not been initialize, call `begin_epoch`".to_string())
+        })?;
+        if self.sink_writer.is_none() {
+            self.create_sink_writer(epoch).await?;
+        }
+        if self.is_append_only {
+            self.append_only(chunk).await
+        } else {
+            // currently file sink only supports append only mode.
+            unimplemented!()
+        }
+    }
+
+    async fn begin_epoch(&mut self, epoch: u64) -> Result<()> {
+        self.epoch = Some(epoch);
+        Ok(())
+    }
+
+    async fn abort(&mut self) -> Result<()> {
+        Ok(())
+    }
```

Comment:
There is no need to implement this method here because we didn't change the default implementation.